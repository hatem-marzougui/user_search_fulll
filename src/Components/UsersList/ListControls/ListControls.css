.control-div {
  border: 1px solid #ccc;
  height: 25px;
  display: flex;
  justify-content: space-between;
  align-items: center;
  padding: 5px 30px;
}
.icon-button {
  background: none;
  border: none;
  padding: 4px;
  cursor: pointer;
}
.icon {
  margin-right: 5px;
}
.users-count {
  display: flex;
  align-items: center;
  justify-content: center;
}
.select-all-checkbox {
  position: absolute;
  top: 0;
  left: 0;
  margin-right: 10px;
  cursor: pointer;
  width: 18px;
  height: 18px;
  position: absolute;
  top: 2px;
  left: 5px;
  border-radius: 5px;
}

/* The switch - the box around the slider */
.switch {
  position: relative;
  display: inline-block;
  width: 40px;
  height: 20px;
  margin-right: 10px;
}

/* Hide default HTML checkbox */
.switch input {
  opacity: 0;
  width: 0;
  height: 0;
}

/* The slider */
.slider {
  position: absolute;
  cursor: pointer;
  top: 0;
  left: 0;
  right: 0;
  bottom: 0;
  background-color: #ccc;
  -webkit-transition: 0.4s;
  transition: 0.4s;
}

.slider:before {
  position: absolute;
  content: "";
  height: 20px;
  width: 20px;
  left: 0px;
  bottom: 0px;
  background-color: white;
  -webkit-transition: 0.4s;
  transition: 0.4s;
}

input:checked + .slider {
  background-color: #2196f3;
}

input:focus + .slider {
  box-shadow: 0 0 1px #2196f3;
}

input:checked + .slider:before {
  -webkit-transform: translateX(20px);
  -ms-transform: translateX(20px);
  transform: translateX(20px);
}

/* Rounded sliders */
.slider.round {
  border-radius: 34px;
}

.slider.round:before {
  border-radius: 50%;
}
.edit-mode-toggle {
  position: absolute;
  right: 2rem;
}

.controls-buttons {
  position: absolute;
  right: 12rem;
  display: flex;
  justify-content: center;
  align-items: center;
}
/* ðŸ“± Jusquâ€™Ã  480px â€” hide text number of elements selected */
@media (max-width: 480px) {
  .users-count {
    display: none;
  }
}
