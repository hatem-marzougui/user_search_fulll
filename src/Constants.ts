export const HEADER_TITLE: string = "Github Search";
export const VIEW_PROFILE_BUTTON_TEXT: string = "View Profile";
export const LOGIN: string = "Login";
export const EDIT_MODE: string = "Edit Mode";
export const DUPLICATE_USERS: string = "Duplicate Users";
export const SEARCH_PLACEHOLDER: string = "Search...";
export const USERS_API_ENDPOINT: string = "/search/users";
export const RATE_LIMIT_ERROR_MESSAGE: string =
  "GitHub request limit reached. Please try again in a minute.";
export const EMPTY_SEARCH_QUERY_ERROR: string = "Please enter a search query";
export const NO_RESULTS_FOUND: string = "No result found  ";
export const USER_CARD_AVATAR_ALT: string = "User Avatar";
export const USER_CARD_ID_LABEL: string = "ID:";
export const USER_CARD_LOGIN_LABEL: string = "Login:";
export const USER_CARD_VIEW_PROFILE_BUTTON_TEXT: string = "View Profile";
export const RATE_LIMIT: string = "rate limit";
export const ELEMENT_SELECTED: string = "elements selected";
