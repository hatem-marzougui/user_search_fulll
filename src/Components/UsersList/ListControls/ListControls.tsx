import React from "react";
import { EDIT_MODE, ELEMENT_SELECTED } from "../../../Constants";
import "./ListControls.css";
import { User } from "../../../types/user";

type ListControlsProps = {
  selectedItems: string[];
  handleDuplicateUsers: () => void;
  handleDeleteUsers: () => void;
  githubUsers: User[];
  editMode: boolean;
  handleEditMode: (e: React.ChangeEvent<HTMLInputElement>) => void;
};

const ListControls = ({
  selectedItems,
  handleDuplicateUsers,
  handleDeleteUsers,
  githubUsers,
  editMode,
  handleEditMode,
}: ListControlsProps) => {
  return (
    <div className="control-div">
      <div className="users-count">
        <img
          className="icon"
          src="/icons/list.svg"
          alt="liste"
          width={24}
          height={24}
        />
        <span>
          {selectedItems.length} {ELEMENT_SELECTED}
        </span>
      </div>
      {
        <div className="controls-buttons">
          {selectedItems.length > 0 && (
            <>
              <button className="icon-button" onClick={handleDuplicateUsers}>
                <img
                  className="icon"
                  src="/icons/duplicate.svg"
                  alt="duplicate"
                  width={24}
                  height={24}
                />
              </button>
              <button className="icon-button" onClick={handleDeleteUsers}>
                <img
                  className="icon"
                  src="/icons/delete.svg"
                  alt="delete"
                  width={24}
                  height={24}
                />
              </button>
            </>
          )}
          {githubUsers.length > 0 && (
            <>
              <label className="switch">
                <input
                  type="checkbox"
                  checked={editMode}
                  onChange={handleEditMode}
                />
                <span className="slider round"></span>
              </label>
              <span>{EDIT_MODE} </span>
            </>
          )}
        </div>
      }
    </div>
  );
};

export default ListControls;
