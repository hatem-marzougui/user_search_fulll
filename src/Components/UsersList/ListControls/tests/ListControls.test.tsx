import { render, screen, fireEvent } from "@testing-library/react";
import ListControls from "../ListControls";
import { EDIT_MODE, ELEMENT_SELECTED } from "../../../../Constants";

describe("ListControls simple test", () => {
  const mockDuplicate = jest.fn();
  const mockDelete = jest.fn();
  const mockSwitchEditMode = jest.fn();

  test("renders selected count and delete/duplicate buttons when items are selected", () => {
    render(
      <ListControls
        selectedItems={["1"]}
        handleDuplicateUsers={mockDuplicate}
        handleDeleteUsers={mockDelete}
        githubUsers={[{ id: 1, login: "user", avatar_url: "" }]}
        editMode={false}
        handleEditMode={mockSwitchEditMode}
      />
    );
    expect(screen.getByText(`1 ${ELEMENT_SELECTED}`)).toBeInTheDocument();
    fireEvent.click(screen.getByAltText("duplicate"));
    expect(mockDuplicate).toHaveBeenCalled();

    fireEvent.click(screen.getByAltText("delete"));
    expect(mockDelete).toHaveBeenCalled();
  });

  test("renders edit mode switcher", () => {
    render(
      <ListControls
        selectedItems={[]}
        handleDuplicateUsers={mockDuplicate}
        handleDeleteUsers={mockDelete}
        githubUsers={[{ id: 1, login: "user", avatar_url: "" }]}
        editMode={true}
        handleEditMode={mockSwitchEditMode}
      />
    );

    const switcher = screen.getByRole("checkbox");
    expect(switcher).toBeChecked();

    fireEvent.click(switcher);
    expect(mockSwitchEditMode).toHaveBeenCalled();
  });

  test("matches snapshot", () => {
    const { container } = render(
      <ListControls
        selectedItems={[]}
        handleDuplicateUsers={mockDuplicate}
        handleDeleteUsers={mockDelete}
        githubUsers={[{ id: 1, login: "user", avatar_url: "" }]}
        editMode={true}
        handleEditMode={mockSwitchEditMode}
      />
    );
    expect(container).toMatchSnapshot();
  });
});
