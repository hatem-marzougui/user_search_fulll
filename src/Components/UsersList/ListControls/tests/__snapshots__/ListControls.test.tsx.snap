// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ListControls simple test matches snapshot 1`] = `
<div>
  <div
    class="control-div"
  >
    <div
      class="users-count"
    >
      <img
        alt="liste"
        class="icon"
        height="24"
        src="/icons/list.svg"
        width="24"
      />
      <span>
        0
         
        elements selected
      </span>
    </div>
    <div
      class="controls-buttons"
    >
      <label
        class="switch"
      >
        <input
          checked=""
          type="checkbox"
        />
        <span
          class="slider round"
        />
      </label>
      <span>
        Edit Mode
         
      </span>
    </div>
  </div>
</div>
`;
